fuse/dl/models/backbones/backbone_transformer.py

```diff
@@ -36,12 +38,146 @@ def __init__(
+        -> Apply linear layer on the last encoded vector
+
+    Building Blocks:
+        In this architecture we use three components from "x-transformers" library.
+        Here is a short summary - for more information consider check source code.
+        * TransformerWrapper - wraps an attention layer (in our case the Encoder) and applies token & positional embedding.
+        * Encoder - self attention layers. In our case it gets the embedding from the wrapper.
+        * CrossAttender - cross attention layers. In our case it gets the embedding from the encoders.
+
+    (1) see the following blog post for more info regarding cross attention in transformers:
+        https://vaclavkosar.com/ml/cross-attention-in-transformer-architecture
+    """
+
+    def __init__(
+        self,
+        emb_dim: int,
+        num_tokens_a: int,
+        num_tokens_b: int,
+        max_seq_len_a: int,
+        max_seq_len_b: int,
+        depth_a: int = 6,
+        depth_b: int = 6,
+        depth_cross_attn: int = 6,
+        heads_a: int = 9,
+        heads_b: int = 9,
+        output_dim: Optional[int] = None,
+        context: str = "seq_b",
+        kwargs_wrapper_a: Optional[dict] = None,
+        kwargs_wrapper_b: Optional[dict] = None,
+        kwargs_encoder_a: Optional[dict] = None,
+        kwargs_encoder_b: Optional[dict] = None,
+        kwargs_cross_attn: Optional[dict] = None,
+    ):
+        """
+        :param emb_dim: inner model dimension
+        :param num_tokens_a: number of tokens of the first sequence
+        :param num_tokens_b: number of tokens of the second sequence
+        :param max_seq_len_a: the maximum length of the first sequence
+        :param max_seq_len_b: the maximum length of the second sequence
+        :param depth_a: first sequence encoder's depth
+        :param depth_b: second sequence encoder's depth
+        :param depth_cross_attn: cross attender(s)' length
+        :param heads_a: number of attention heads for the first sequence's encoder
+        :param heads_b: number of attention heads for the second sequence's encoder
+        :param output_dim: (optional) model's output dimension. if not give the emb dim will be used as default.
+        :param context: which sequence will be used as context in the cross attention module:
+                        "seq_a": the first sequence will be used as a context
+                        "seq_b": the second sequence will be used as a context
+                        "both": will use two cross attention modules to take each one of the sequences as a context to the other one.
+        """
+        super().__init__()
+
+        if output_dim is None:
+            output_dim = emb_dim
+
+        assert context in ["seq_a", "seq_b", "both"]
+        self._context = context
+
+        # init sequences' encoders
+        self.enc_a = TransformerWrapper(
+            num_tokens=num_tokens_a,
+            max_seq_len=max_seq_len_a,
+            **kwargs_wrapper_a,
```

Comment:
You marked kwargs_wrapper_a as optional - but it seems like you'll get an error if you don't set it.

Reply:
nice catch! missed that, thanks!